portal-impl/src/com/liferay/portal/tools/DBUpgrader.java

```diff
@@ -195,6 +214,108 @@ public static void verify() throws VerifyException {
 		verifyResourcePermissions.verify();
 	}
 
+	public static void waitForLocks() throws Exception {
+		if (!PropsValues.UPGRADE_DATABASE_MANAGED_STARTUP) {
+			return;
+		}
+
+		while (true) {
+			if (!_hasLockTable()) {
+				return;
+			}
+
+			try (Connection connection = DataAccess.getConnection();
+				PreparedStatement preparedStatement =
+					_getSelectLockPreparedStatement(connection);
+				ResultSet resultSet = preparedStatement.executeQuery()) {
+
+				if (!resultSet.next()) {
+					return;
+				}
+			}
+			catch (SQLException sqlException) {
+				if (_log.isDebugEnabled()) {
+					_log.debug(sqlException);
+				}
+			}
+
+			Thread.sleep(PropsValues.UPGRADE_DATABASE_LOCK_REFRESH_TIME);
+		}
+	}
+
+	private static Lock _acquireLock() throws Exception {
+		_createLockTable();
+
+		long lockId;
+
+		while (true) {
+			if (!_hasLockTable()) {
+				return null;
+			}
+
+			try (Connection connection = DataAccess.getConnection();
+				PreparedStatement preparedStatement =
+					_getSelectLockPreparedStatement(connection);
+				ResultSet resultSet = preparedStatement.executeQuery()) {
+
+				if (resultSet.next()) {
+					lockId = resultSet.getLong(1);
+
+					break;
+				}
+
+				lockId = SecureRandomUtil.nextLong();
+
+				try (PreparedStatement preparedStatement2 =
+						_getInsertLockPreparedStatement(connection, lockId)) {
+
+					if (preparedStatement2.executeUpdate() == 1) {
+						break;
+					}
+				}
+				catch (SQLException sqlException) {
+					if (_log.isDebugEnabled()) {
+						_log.debug(sqlException);
+					}
+				}
+			}
+		}
+
+		Connection connection = DataAccess.getConnection();
+
+		boolean autoCommit = connection.getAutoCommit();
+
+		while (true) {
+			if (!_hasLockTable()) {
+				connection.setAutoCommit(autoCommit);
+				connection.close();
+
+				return null;
+			}
+
+			try (PreparedStatement preparedStatement =
```

Comment:
@LuisOrtiz89 this is fundamentally wrong here. This solution is building on top of "select for update" row lock. When multiple portal instances fighting to execute the upgrade, they need to compete the row lock in order to decide who runs the upgrade, who wait for the upgrade to be done.

The execution of "select for update" statement opens a protection domain, once entered, the current node knows it is in mutual exclusive mode. It needs to exam whether the upgrade is needed, if yes, do it, if no, release the row lock to continue startup.

In order to keep the protection domain, the "select for update" statement's connection's tx can not be committed, once commit it, the lock will be released. But here you always commit the tx and close the connection, this "lock" is not protecting anything.